packages/ai-jsx/src/core/constrained-output.tsx

```diff
@@ -0,0 +1,141 @@
+/**
+ * This module defines affordances for constraining the output of the model
+ * into specific formats, such as JSON, YAML, or Markdown.
+ */
+import * as LLMx from '../index.js';
+import { ChatCompletion, SystemMessage, AssistantMessage, UserMessage } from './completion';
+import { memo } from './memoize';
+import yaml from 'js-yaml';
+
+function isJsonString(str: string): boolean | string {
+  try {
+    JSON.parse(str);
+  } catch (e: any) {
+    return e.message;
+  }
+  return true;
+}
+
+function isYamlString(str: string): boolean | string {
+  try {
+    // try to parse the string as YAML
+    yaml.load(str);
+  } catch (e: any) {
+    return e.message;
+  }
+  return true;
+}
+
+/**
+ * ChatCompletion that constrains the output to be a valid object format (e.g. JSON/YAML).
+ *
+ *
+ * @param type_name name of the type, to be used in the prompt
+ * @param validator function that returns true if the output is valid, or a string with an error message if not
+ *     // TODO: what should the signature of this function be?
+ * @param retries if output fails validation, how many times should we retry?
+ * @param verbose returns intermediate output if true
+ * @returns a string that validates as the given type (TODO: should I return the parsed object instead?)
+ */
+async function* ObjectFormatChatCompletion(
+  {
+    retries = 2,
+    verbose = false,
```

Comment:
Don't take a verbosity param. Just use the log levels. `logger.trace`, `debug`, `warn`, etc.

Reply:
Oh I was hoping there would be something like this.

Reply:
Actually, I agree for some lines, but for some, the idea was that if I `render` something internally, you can't see it anywhere even with `showInspector`.
This might be solved with better observability, so maybe I shouldn't worry about it.

Reply:
Yeah, `showInspector` has a gap right now where it doesn't get those internal `render`s, as you point out. We just need to find a better way to do that.